File: FindMinRotatedSortedArray.java
// Time Complexity : O(logn)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

class Solution {
    public int findMin(int[] nums) {
        if(nums == null || nums.length == 0)
            return -1;
        
        int low = 0;
        int high = nums.length - 1;
        
        while(low <= high){
            if(nums[low] <= nums[high])
                return nums[low];
            
            int mid = low + (high - low)/2;
            
            if((mid == low || nums[mid - 1] > nums[mid]) && (mid == high || nums[mid + 1] > nums[mid]) )
                return nums[mid];
            else if(nums[low] <= nums[mid])
                low = mid + 1;
            else
                high = mid - 1;
        }
        return -1;
    }
}

File: find_min_rotated_sorted_array.cpp
// Time Complexity : O(logn)
// Space Complexity :  O(n) - In Place
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this 

class Solution {
public:
    
    int findPivot(vector<int>& nums, int l, int r)
    {
        // Pivot not found condition
        //Implies an unrotated array
        if(l == r)
        {
            return -1;
        }
        
        int m = floor((l+r)/2);
        
        // Found the largest element
        if(nums[m] > nums[m+1])
        {
            return m;
        }
        // Left Half condition
        else if(nums[l] > nums[m])
        {
            return findPivot(nums, l, m);
        }
        // Right Half condition
        else if(nums[m+1] > nums[r])
        {
            return findPivot(nums, m+1, r);
        }
        else
        {
            cout << "Shouldn't come here";
            return -1;
        }
    }
    
    int findMin(vector<int>& nums) {
        
        int idx = findPivot(nums, 0, nums.size()-1);

        //idx+1 because Smallest element needed      
        return nums[idx+1];
        
    }
};